dvc/remote/gdrive/client.py

```diff
@@ -0,0 +1,284 @@
+from time import sleep
+import logging
+import posixpath
+import os
+
+from funcy import cached_property
+
+from requests import ConnectionError
+
+from dvc.progress import progress
+from dvc.remote.gdrive.utils import (
+    metadata_isdir,
+    response_is_ratelimit,
+    MIME_GOOGLE_APPS_FOLDER,
+    response_error_message,
+)
+from dvc.remote.gdrive.exceptions import (
+    GDriveError,
+    GDriveHTTPError,
+    GDriveResourceNotFound,
+)
+from dvc.remote.gdrive.oauth2 import OAuth2
+
+logger = logging.getLogger(__name__)
+
+
+class GDriveClient:
+
+    GOOGLEAPIS_BASE_URL = "https://www.googleapis.com/"
+    TIMEOUT = (5, 60)
+
+    def __init__(
+        self,
+        space,
+        oauth_id,
+        credentialpath,
+        scopes,
+        oauth2_flow_runner,
+        max_retries=10,
+    ):
+        self.space = space
+        self.oauth_id = oauth_id
+        self.credentialpath = credentialpath
+        self.scopes = scopes
+        self.oauth2_flow_runner = oauth2_flow_runner
+        self.max_retries = max_retries
+        self.oauth2 = OAuth2(
+            oauth_id, credentialpath, scopes, oauth2_flow_runner
+        )
+
+    @cached_property
+    def session(self):
+        """AuthorizedSession to communicate with https://googleapis.com
+
+        Security notice:
+
+        It always adds the Authorization header to the requests, not paying
+        attention is request is for googleapis.com or not. It is just how
+        AuthorizedSession from google-auth implements adding its headers. Don't
+        use RemoteGDrive.session() to send requests to domains other than
+        googleapis.com.
+        """
+        return self.oauth2.get_session()
+
+    def request(self, method, path, *args, **kwargs):
+        # Google Drive has tight rate limits, which strikes the
+        # performance and gives the 403 and 429 errors.
+        # See https://developers.google.com/drive/api/v3/handle-errors
+        retries = 0
+        exponential_backoff = 1
+        if "timeout" not in kwargs:
+            kwargs["timeout"] = self.TIMEOUT
+        while retries < self.max_retries:
+            retries += 1
+            response = self.session.request(
+                method, self.GOOGLEAPIS_BASE_URL + path, *args, **kwargs
+            )
+            if response_is_ratelimit(response) or response.status_code >= 500:
```

Comment:
does it return some header that specifies how long should we wait before retrying?

Reply:
Nope :(. As far as I could recognize.

Reply:
They just suggest to use exponential backoff technics.

Reply:
👍